### PR TITLE
drivers: nxp_enet: Enable on k6x and RT11xx

### PR DESCRIPTION
**Pull request by decsny**
Enable NXP ENET driver for the freedom k64f board and RT11xx

### CODE REVIEW

File: drivers/ethernet/eth_nxp_enet.c [hunk 1]
```diff
@@ -1040,4 +1040,29 @@ static const struct ethernet_api api_funcs = {
...
+{											\
+	clock_control_on(DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(n)),				\
+			(void *)DT_INST_CLOCKS_CELL_BY_IDX(n, 0, name));		\
+											\
+	ENET_Reset((ENET_Type *)DT_INST_REG_ADDR(n));					\
+											\
+	return 0;									\
+}											\
+											\
+	/* Init the module before any of the MAC, MDIO, or PTP clock */			\
```
Comment (dleach02) [inline drivers/ethernet/eth_nxp_enet.c:1065]: Your macro continuation '\' went too far

File: drivers/ethernet/eth_nxp_enet.c [hunk 2]
```diff
@@ -791,7 +791,7 @@ static int eth_nxp_enet_init(const struct device *dev)
 	enet_config.callback = eth_callback;
 	enet_config.userData = (void *)dev;
 
-	ENET_Init(config->base,
+	ENET_Up(config->base,
```
[the remark this answers is not in the document]
Comment (decsny) [reply drivers/ethernet/eth_nxp_enet.c:794]: I am not sure, but with this target, I noticed the debugger could not read the ENET registers during mdio initialization, and this change fixed that issue

File: drivers/ethernet/eth_nxp_enet.c (hunk 1 above)
Comment (dleach02) [reply drivers/ethernet/eth_nxp_enet.c:1065]: ah yes, I missed that

File: boards/arm/frdm_k64f/doc/index.rst [hunk 3]
```diff
@@ -378,3 +378,13 @@ of pyocd commands:
...
+
+Experimental ENET Driver
+========================
+
+Current default ethernet driver is eth_mcux, with binding `nxp,kinetis-ethernet`. There is a new
+driver with binding `nxp,enet`, which is experimental and undergoing development, but will have
+enhanced capability, such as not hardcoding code for only one phy in the driver like eth_mcux.
+
+To build for this EVK with the new driver, include the experimental overlay to west build with
+the option `-DEXTRA_DTC_OVERLAY_FILE=nxp,enet-experimental.overlay`.
```
Comment (hakehuang) [inline boards/arm/frdm_k64f/doc/index.rst:390]: can we add this option as a new test case in one of the network testing sample.yaml file

Comment (decsny) [reply boards/arm/frdm_k64f/doc/index.rst:390]: ok

**Review by dleach02 — DISMISSED** [review]

File: drivers/ethernet/eth_nxp_enet.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (hakehuang) [reply drivers/ethernet/eth_nxp_enet.c:803]: we do not call ENET_Init anymore, which would be dangous if SDK changes.

**Comment by hakehuang** [comment]
test with zperf UDP, no issue, but I find the link detect does not work, the link status keeps on by default, no matter I plug in or plug our the ethernet cable. is this supported?
```
uart:~$ zperf udp upload 192.0.2.2 5001 10 1K 1000M
Remote port is 5001
Invalid IPv6 address 192.0.2.2
Connecting to 192.0.2.2
Duration:       10.00 s
Packet size:    1024 bytes
Rate:           1024000 kbps
Starting...
Rate:           1000.00 Mbps
Packet duration 7 us
-
Upload completed!
Statistics:             server  (client)
Duration:               10.00 s (10.00 s)
Num packets:            28752   (28752)
Num packets out order:  0
Num packets lost:       0
Jitter:                 45 us
Rate:                   22.46 Mbps      (22.46 Mbps)
```

**Comment by decsny** [comment]
> @decsny test with zperf UDP, no issue, but I find the link detect does not work, the link status keeps on by default, no matter I plug in or plug our the ethernet cable. is this supported?

are you seeing this only on the k64?

**Comment by decsny** [comment]
> @decsny test with zperf UDP, no issue, but I find the link detect does not work, the link status keeps on by default, no matter I plug in or plug our the ethernet cable. is this supported?

okay, I see myself now too there is some issue with this on the kinetis board, which doesn't exist on the RT boards. thanks for reporting this

File: drivers/ethernet/eth_nxp_enet.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (dleach02) [reply drivers/ethernet/eth_nxp_enet.c:803]: Good point. We use the ENET_Init() to deal with platform specific stuff.

**Comment by decsny** [comment]
@hakehuang I think I fixed the issues in the latest push, can you check

**Comment by hakehuang** [comment]
> @hakehuang I think I fixed the issues in the latest push, can you check

@decsny , see below log, although the phy error is reported, but the link status is still on, although I did not plugin the ethernet cable, and after that no even triggered with cable in / out.

```
[00:00:04.553,000] <err> phy_mc_ksz8081: Error in autonegotiation
*** Booting Zephyr OS build zephyr-v3.5.0-2749-g903783a86f7d ***
[00:00:04.555,000] <inf> net_config: Initializing network
[00:00:04.555,000] <inf> net_config: IPv4 address: 192.0.2.1
[00:00:04.655,000] <inf> net_config: IPv6 address: 2001:db8::1
uart:~$ ifconfig
ifconfig: command not found
uart:~$ net iface

Interface eth0 (0x200012a4) (Ethernet) [1]
===================================
Link addr : 02:04:9F:51:82:96
MTU       : 1500
Flags     : AUTO_START,IPv4,IPv6
Device    : ethernet (0x25a70)
Ethernet capabilities supported:
        Virtual LAN
        10 Mbits
        100 Mbits
IPv6 unicast addresses (max 4):
        fe80::4:9fff:fe51:8296 autoconf preferred infinite
        2001:db8::1 manual preferred infinite
IPv6 multicast addresses (max 5):
        ff02::1
        ff02::1:ff51:8296
        ff02::1:ff00:1
IPv6 prefixes (max 2):
        <none>
IPv6 hop limit           : 64
IPv6 base reachable time : 30000
IPv6 reachable time      : 35882
IPv6 retransmit timer    : 0
IPv4 unicast addresses (max 1):
        192.0.2.1 manual preferred infinite
IPv4 multicast addresses (max 2):
        224.0.0.1
IPv4 gateway : 0.0.0.0
IPv4 netmask : 255.255.255.0
```

**Comment by decsny** [comment]
> > @hakehuang I think I fixed the issues in the latest push, can you check
> 
> @decsny , see below log, although the phy error is reported, but the link status is still on, although I did not plugin the ethernet cable, and after that no even triggered with cable in / out.
> 
> ```
> [00:00:04.553,000] <err> phy_mc_ksz8081: Error in autonegotiation
> *** Booting Zephyr OS build zephyr-v3.5.0-2749-g903783a86f7d ***
> [00:00:04.555,000] <inf> net_config: Initializing network
> [00:00:04.555,000] <inf> net_config: IPv4 address: 192.0.2.1
> [00:00:04.655,000] <inf> net_config: IPv6 address: 2001:db8::1
> uart:~$ ifconfig
> ifconfig: command not found
> uart:~$ net iface
> 
> Interface eth0 (0x200012a4) (Ethernet) [1]
> ===================================
> Link addr : 02:04:9F:51:82:96
> MTU       : 1500
> Flags     : AUTO_START,IPv4,IPv6
> Device    : ethernet (0x25a70)
> Ethernet capabilities supported:
>         Virtual LAN
>         10 Mbits
>         100 Mbits
> IPv6 unicast addresses (max 4):
>         fe80::4:9fff:fe51:8296 autoconf preferred infinite
>         2001:db8::1 manual preferred infinite
> IPv6 multicast addresses (max 5):
>         ff02::1
>         ff02::1:ff51:8296
>         ff02::1:ff00:1
> IPv6 prefixes (max 2):
>         <none>
> IPv6 hop limit           : 64
> IPv6 base reachable time : 30000
> IPv6 reachable time      : 35882
> IPv6 retransmit timer    : 0
> IPv4 unicast addresses (max 1):
>         192.0.2.1 manual preferred infinite
> IPv4 multicast addresses (max 2):
>         224.0.0.1
> IPv4 gateway : 0.0.0.0
> IPv4 netmask : 255.255.255.0
> ```

can you clarify what part of this log is the problem ?

**Comment by hakehuang** [comment]
> can you clarify what part of this log is the problem ?

@decsny , as I do not plugin the eth-cable, the link status should be none,  but it shows linked up. also, there used to be a LOG_INF to show the link status change, in eth_mcux, which is helpful usually for user to check the status. and it is gone.

**Comment by decsny** [comment]
@hakehuang There was a logging message about this from the PHY driver, but I also added one to the MAC driver, if you think it is useful.

**Comment by hakehuang** [comment]
test with frdm_k64f and mimxrt1060_evk, zperf works fine

**Review by hakehuang — APPROVED** [review]

**Review by danieldegrasse — APPROVED** [review]

File: drivers/ethernet/eth_nxp_enet.c [hunk 4]
```diff
@@ -1047,4 +1047,29 @@ static const struct ethernet_api api_funcs = {
...
+	clock_control_on(DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(n)),				\
+			(void *)DT_INST_CLOCKS_CELL_BY_IDX(n, 0, name));		\
+											\
+	ENET_Reset((ENET_Type *)DT_INST_REG_ADDR(n));					\
+											\
+	return 0;									\
+}											\
+											\
+	/* Init the module before any of the MAC, MDIO, or PTP clock */			\
+	SYS_INIT(nxp_enet_##n##_init, POST_KERNEL, 0);
```
Comment (mmahadevan108) [inline drivers/ethernet/eth_nxp_enet.c:1075]: Any reason this is tabbed in? Since it is not part of the function maybe the tab is not needed.

Comment (decsny) [reply drivers/ethernet/eth_nxp_enet.c:1075]: it's part of the macro function

File: boards/arm/frdm_k64f/doc/index.rst (hunk 3 above)
Comment (thedjnK) [inline boards/arm/frdm_k64f/doc/index.rst:390]: seems like this should be made into a snippet?

**Pull request merged**